tests/CMakeLists.txt

```diff
@@ -0,0 +1,248 @@
+cmake_minimum_required(VERSION 3.16.0)
```

Comment:
Please remove all tests

Reply:
This is probably the most ridiculous request I've ever heard. (Given the context).

I can't think of any single benefit and the downside is just huge.

The only reason I can think of to remove the tests is something stupid and unprofessional like "we don't want too many files in our project", which makes no sense given a git repository and a modern filesystem.

I can only think of 2 categories of answers:
1. stupid / unprofessional / incompetent
2. malicious / hidden agenda

The tests don't affect anyone negatively or break anything or interfere with anything. They only bring in benefits. (`stubs.cpp` is the exception, but the size of the file is the fault of Domoticz being 100% monolithic).

So I assume there's some serious misunderstanding here.

All other objections have been handled:
- they don't interfere with anything (separate folder / 100% independent build step)
- they test mostly the MainWorker and SQLHelper (so they're 100% dependent on Domoticz)
- they assure a level of quality (so you'd have to hate users and developers to have a reason to remove them)
- they're well written, well organized and generic (so they're useful for other JSON/HTTP based integrations)
- it's an industry standard to have at least basic unit tests included in quality open source projects
- the maintenance cost is ZERO - if they stop working, you can ignore them until someone sends a PR to fix them
- they build pretty fast (given that precompiled headers and poorly organized in Domoticz itself, which is why the main build is so slow to begin with)
- they run under 300 ms on Travis

I could go on...

Here is an example of them running on Travis (after a main build):

https://travis-ci.org/domoticz/domoticz/builds/647552213?utm_source=github_status&utm_medium=notification

Rejecting tests like these in this situation seems like something someone posted on: https://thedailywtf.com/

Especially since you have the option to include the tests and then just ignore them. (So whoever wants to use them can).

And you can always ping me to help out anyone who'd want to use (or fix) these tests.

Reply:
Have you looked at the dzVents tests? There are almost 1000 tests including many integration tests. Imo it's always good to have tests and I strongly believe Domoticz should have more of them. Without these dzVents tests it wouldn't be possible for me to maintain the project (dzVents). It happened more than once that someone else changed something in Domoticz core causing essential parts to stop functioning. The tests revealed this.

Reply:
Yes, the dzVents tests were the first thing I tried to use.

While these seem useful for maintainers, they have 2 major flaws:
1. The unit tests obviously are pretty much useless for integration developers
2. The integration tests require a running Domoticz instance

The second is particularly problematic for developing c++ integrations:
- extremely long compile times (due to broken precompiled header support, which this PR allows fixing) to rerun tests
- difficult to debug segfaults and trace Domoticz internals - inconvenient to wait for build, server startup, and doing all those steps manually every time
- difficult to setup edge cases in database
- no way to drive/control the main worker (needed for edge cases in integration tests)
- not suitable for TDD (small changes in code + immediate feedback), which is crucial for development and refactoring in Domoticz (otherwise development is extremely discouraging)
- difficult to reliably run server instances for tests on Travis (which is probably why the existing dzVents are failing - 4 errors broken in upstream/development)
- testing some edge cases may require additional test-specific functionality exposed via Lua bindings (which would mean more changes to the core, which are beyond the scope of this PR)
- the tests are already really slow (even when run against a compiled domoticz) and that means the more tests are added (for edge cases), the slower the tests become (until no one will want to run them)

Meanwhile, these tests:
- are easy to add to Travis
- they bulid really fast (and can build even faster if tweaked), even on Travis
- they run under 300 milliseconds, even on Travis
- they don't require any new dependencies (except Google test/mock as a submodule, which is the recommended way and has no downside)
- they assure the tested integrations work, even without requiring actual devices for tests
- they don't require any changes in the core code to work (no extra bridges or layers)
- Google mock allows conveniently stubbing/mocking out any functionality (e.g. HTTP requests - which isn't possible with Lua integration tests)
- the tests allowed me to quickly migrate away from pTypeLighting2 to pGeneralDevice without debugging (which was necessary to work around a pTypeLighting2 bug)
- the tests were a great way to understand how Domoticz MainWorker handles each device type, without having to debug through the code full of gigantic switch statements and vendor-specific workarounds and hacks
- the tests support 11 new, very different devices through integration tests (meaning, using the database and MainWorker just as it does normally), and these would be very difficult to port to the Lua tests (which would make the tests likely umaintainable and would require lots of code changes in Domoticz). This alone is a huge achievement that would help a lot of device integrators while allowing the project owners to change anything they want in Domoticz internals and release confidently knowing they didn't break anything. Also, those 11 devices have almost identical tests (except for actual values and naming), which shows how generic and reusable they are. (And how difficult it is to develop reliable integrations for Domoticz - which is why these tests are crucial).
- these tests are heavily tied to Domoticz (calling the worker through the same API the UI ends up calling, and comparing Domoticz-specific database states with expectations), so they they can't even be run without a Domoticz source tree - and it makes zero sense to clone Domoticz, patch it, add tests and run this in a separate github/travis project ... just to "not have it in Domoticz". And those tests are really just testing to make sure Domoticz isn't broken - while testing on Dummy devices doesn't expose all the real-life edge cases that happen with real devices.
- code coverage is available to show device integration coverage (which is really, really high right now) and it highlights the paths takes in MainWorker/SQLHelper code. Adding code coverage to the whole codebase while running lua integration tests would be extremely slow and add a lot of noise (lots of paths triggered during test setup and app startup). Also, the reports would be huge.
- the tests are compatible with unity builds (I used Cotire until switching to recent cmake that superseded it), so going from a clean checkout to having test results and coverage reports is really fast and automatic. 
- many popular and respected projects EXPECT unit tests to be added along with changes - precisely to reduce testing and maintenance burden
- Lua tests don't have compile-time type checking, which is extremely useful for a C++ project like Domoticz, where defensive coding and type safety can prevent bugs from being introduce in code to begin with
- Lua tests aren't testing any specific integration (except for dummy devices), so they're not good, proven examples to use as a base for testing new/existing integrations, while the coverage in these tests show that ANYONE can make changes in Domoticz or the integrations covered by these tests and have confidence they didn't break anything BEFORE even submitting a PR. Again, the automation, convenience and speed of such tests is crucial to get people using them. (And adding tests themselves).
- the tests are very well designed, with the separation of hardware and "discoverable" features, and it makes the best usage of Domoticz architecture, so any new integrations that fit well with these exact tests with have a decent level of quality to start of with. This also means new devices can be added - along with their tests - mostly by copy&paste (or making things generic and introducing them into Domoticz). Otherwise, other developers new to Domoticz will copy whatever they pick, which means lots of bad practices, using outdated APIs and approaches, lots of confusion and lots of rework (and often after they've already invested lots of time manually testing their work for a PR). A single, maintained unit test in C++ like these would've saved me a lot of time and confusion.
- it's a separate folder requiring no changes to Domoticz itself (which is a technical miracle, considering how monolithic Domoticz is - you cannot otherwise build a binary for testing without including EVERYTHING except `main()` from `main/domoticz.cpp`). 

Also, Domoticz has an incredibly steep learning curve - which means lots of human lives wasted (instead of used for more valuable work, like adding useful features) just to avoid "having too many files in the project". Unit tests at least show HOW to do such things in a RECOMMENDED way. Throwing this away is like deliberately making development hard and inconvenient to frustrate and discourage people on purpose.

I've addressed all the concerns I could think of (most addressed above).

If you really don't want "too many files", I can just glue the test files into one giant file - which would be just as stupid to do. (And more inconvenient for everyone with no upside).

Since I've explained and covered so much about these tests (and why the lua tests aren't sufficient at all), I assume there's no reasonable reason to reject THESE tests in THIS situation.